proto/asn1-pdu/x509_certificate_to_der.cc

```diff
@@ -51,30 +51,53 @@ DECLARE_ENCODE_FUNCTION(ExtendedKeyUsage) {
 }
 
 DECLARE_ENCODE_FUNCTION(BasicConstraints) {
+  // RFC 5280, 4.2.1.9: |BasicConstraints| is a sequence of |ca| and
+  // |path_len_constraint|. Save the current size in |tag_len_pos| to place
+  // sequence tag and length after the values are encoded.
+  size_t tag_len_pos = der.size();
+
+  // RFC 5280, 4.2.1.9: |ca| is BOOLEAN DEFAULT FALSE.
+  // (X.690 (2015), 11.5): DEFAULT value in a sequence field is not encoded.
   if (val.ca().val()) {
     Encode(val.ca(), der);
   }
+
+  // RFC 5280, 4.2.1.9: |path_len_constrant| is OPTIONAL. Encode only if
+  // present.
   if (val.has_path_len_constraint()) {
     Encode(val.path_len_constraint(), der);
   }
+
+  // The current size of |der| subtracted by |tag_len_pos|
+  // equates to the size of the sequence |BasicConstraints|.
+  EncodeTagAndLength(kAsn1Sequence, der.size() - tag_len_pos, tag_len_pos, der);
 }
 
 DECLARE_ENCODE_FUNCTION(KeyUsage) {
+  constexpr struct Mask {
+    using Fn = bool (KeyUsage::*)(void) const;
+    Fn use_mask;
+    uint32_t mask;
```

Comment:
Sorry, my naming was so bad, I should have done better for a suggestion.

```suggestion
    Fn key_usage_set_fn;
    uint32_t key_usage_value;
```

I'm not sure that's much better, but I'm specifically trying to make lines 93-96 be a bit more descriptive. I think abbreviating "function" to "fn" here should be fine (in the variable name), but I think the goal is, C++ syntax notwithstanding, can you read line 93-96 and make sense of what's going to happen?

Reply:
I agree with you on all your points! Just wondering here if the name `key_usage_use_value` makes more sense here.

Yes, we are setting the function. But when you invoke the function in the for loop, it feels weird to call `key_usage_set_fn` since we aren't setting anything and rather trying to determine whether we should use the `key_usage_value`.

Let me know if I'm missing something here! :)

Reply:
As the name of the function? That feels a bit weird because the bool doesn't return the value. I had been thinking of the function name in the context of what it returns, and sort of

```
if (foo.key_usage_set()) {
  bar |= foo.key_usage_value
}
```

as the construct we're trying to mirror here. Although I suppose

```
if (foo.is_usage_set()) {
  bar |= foo.usage_value;
}
```

could also work?

Reply:
Ah, I see what you mean now by the naming `key_usage_set`. That makes sense, and I prefer that.